circonus-gometrics.go

```diff
@@ -117,57 +129,73 @@ func NewCirconusMetrics() *CirconusMetrics {
 	}
 
 	return &CirconusMetrics{
-		InstanceId:    fmt.Sprintf("%s:%s", hn, an),
-		SearchTag:     fmt.Sprintf("service:%s", an),
-		ApiHost:       defaultApiHost,
-		ApiApp:        defaultApiApp,
-		Interval:      defaultInterval,
-		Log:           log.New(ioutil.Discard, "", log.LstdFlags),
-		Debug:         false,
-		ready:         false,
-		trapUrl:       "",
-		activeMetrics: make(map[string]bool),
-		counters:      make(map[string]uint64),
-		counterFuncs:  make(map[string]func() uint64),
-		gauges:        make(map[string]int64),
-		gaugeFuncs:    make(map[string]func() int64),
-		histograms:    make(map[string]*Histogram),
-		certPool:      x509.NewCertPool(),
-		checkType:     "httptrap",
+		InstanceId:          fmt.Sprintf("%s:%s", hn, an),
+		SearchTag:           fmt.Sprintf("service:%s", an),
+		ApiHost:             defaultApiHost,
+		ApiApp:              defaultApiApp,
+		Interval:            defaultInterval,
+		MaxSubmissionUrlAge: defaultMaxSubmissionUrlAge,
+		Log:                 log.New(ioutil.Discard, "", log.LstdFlags),
+		Debug:               false,
+		ready:               false,
+		trapUrl:             "",
+		activeMetrics:       make(map[string]bool),
+		counters:            make(map[string]uint64),
+		counterFuncs:        make(map[string]func() uint64),
+		gauges:              make(map[string]int64),
+		gaugeFuncs:          make(map[string]func() int64),
+		histograms:          make(map[string]*Histogram),
+		certPool:            x509.NewCertPool(),
+		checkType:           "httptrap",
 	}
 
 }
 
-// Start starts a perdiodic submission process of all metrics collected
-func (m *CirconusMetrics) Start() {
-	go func() {
-		if m.Debug {
-			m.Log = log.New(os.Stderr, "", log.LstdFlags)
-		}
-		//m.loadCACert()
-		if !m.ready {
-			m.initializeTrap()
+// Start initializes the CirconusMetrics instance based on
+// configuration settings and sets the httptrap check url to
+// which metrics should be sent. It then starts a perdiodic
+// submission process of all metrics collected.
+func (m *CirconusMetrics) Start() error {
+	if m.Debug {
+		m.Log = log.New(os.Stderr, "", log.LstdFlags)
+	}
+	if !m.ready {
+		if err := m.initializeTrap(); err != nil {
+			return err
 		}
```

Comment:
It used to be in a go func but there's no way for the caller to know whether it failed or not during initialization. I'm really not sure which is the better way given what needs to happen for everything to function correctly or to know that it hasn't worked and there will not be metrics. As well as, it will just keep trying every Interval if it did fail which could cause performance issues as well.